### ISSUE TITLE
Investigate different message handling order on PubSub for production and tests

### ISSUE THREAD

**Issue by glevco** [HathorNetwork/hathor-core#552]
> The `cli` import fix is good as is, about the other one I wonder if we should make a test for it.

@jansegre I was investigating this and I found out that existing tests in `test_event_manager.py` were supposed to have caught this. The code in `PubSub` for publishing events is:

```python
    def publish(self, key: HathorEvents, **kwargs: Any) -> None:
        reactor_thread = ReactorThread.get_current_thread(self.reactor)

        args = EventArguments(**kwargs)
        for fn in self._subscribers[key]:
            if reactor_thread == ReactorThread.NOT_RUNNING:
                fn(key, args)
            else:
                is_empty = bool(not self.queue)
                self.queue.append((fn, key, args))
                if is_empty:
                    self._schedule_call_next()
```

The reason the tests passed is because tests use the `Clock` class instead of the reactor, so incidentally things are called in the correct order: the `else` branch is called and the event handler function call is delayed to after the `EventManager` is started. In the real case, when running the full node, the `if reactor_thread == ReactorThread.NOT_RUNNING:` branch is called, and therefore the event handler function is called before the `EventManager` is started.

Also, `ReactorThread` is where the `MAIN_THREAD` value is returned for `Clock` (for testing):

```python
    def get_current_thread(cls, reactor: Reactor) -> 'ReactorThread':
        """ Returns if the code is being run on the reactor thread, if it's running already.
        """
        running = getattr(reactor, 'running', None)
        if running is not None:
            if running:
                return cls.MAIN_THREAD if isInIOThread() else cls.NOT_MAIN_THREAD
            else:
                # if reactor is not running yet, there's no threading
                return cls.NOT_RUNNING
        else:
            # on tests, we use Clock instead of a real Reactor, so there's
            # no threading. We consider that the reactor is running
            return cls.MAIN_THREAD
```

I feel that we should somehow either change `ReactorThread` or `PubSub` themselves, to prevent this from happening again in the future. Code logic is getting executed differently in production and in tests for this reason.

~~For example, changing the above last line from `return cls.MAIN_THREAD` to `return cls.NOT_RUNNING` would fix the problem, but I'm not sure of other unintended consequences.~~ _Edit: I have tested this, and it breaks a lot of other tests. Which may be a bit weird because maybe it indicates that there's code that relies on the specific implementation of PubSub and the order it is going to call message handlers (either at the time of a publish, or the postponed call)..._

What do you think?

_Originally posted by @glevco in https://github.com/HathorNetwork/hathor-core/issues/550#issuecomment-1495280626_

**Comment by glevco** [comment]
This problem is also affecting https://github.com/HathorNetwork/hathor-core/pull/555, so I'm investigating it

**Comment by glevco** [comment]
Another fix has been found to make #555 work, but this problem persists and a solution should be proposed.

**Comment by glevco** [comment]
Is this related? https://github.com/HathorNetwork/hathor-wallet/issues/403
